### PR TITLE
Add __all__ typing to torch.distributed.

### PR DESCRIPTION
**Pull request by kumpera**
Depends on https://github.com/pytorch/pytorch/pull/84119/

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu

### CODE REVIEW

File: torch/distributed/__init__.py [hunk 1]
```diff
@@ -43,8 +41,8 @@ def is_available() -> bool:
         DebugLevel,
         get_debug_level,
         set_debug_level,
-        set_debug_level_from_env,
     )
+    from torch._C._distributed_c10d import set_debug_level_from_env as _set_debug_level_from_env
```
Comment (rohan-varma) [inline torch/distributed/__init__.py:45]: Might be BC breaking, as `set_debug_level_from_env` was public?

File: torch/distributed/__init__.py [hunk 2]
```diff
@@ -67,9 +65,25 @@ def is_available() -> bool:
...
+        rendezvous,
         _create_store_from_options,
     )
 
     from .remote_device import _remote_device
 
-    set_debug_level_from_env()
+    _set_debug_level_from_env()
+
+    def _export_c_types():
```
Comment (rohan-varma) [inline torch/distributed/__init__.py:76]: I'm unfamiliar with this portion of the PR, could you describe it or maybe someone else can help review?

File: torch/distributed/distributed_c10d.py [hunk 3]
```diff
@@ -25,16 +25,32 @@
     Store,
     DebugLevel,
     get_debug_level,
+    Work
```
Comment (rohan-varma) [inline torch/distributed/distributed_c10d.py:28]: is this introducing the work object `dist.Work` as a top-level API, and are we sure we want to do this?

File: torch/distributed/distributed_c10d.py [hunk 4]
```diff
@@ -25,16 +25,32 @@
...
+    'all_to_all_single', 'barrier', 'batch_isend_irecv', 'broadcast',
+    'broadcast_multigpu', 'broadcast_object_list', 'destroy_process_group',
+    'dist_backend', 'gather', 'gather_object', 'get_backend', 'get_rank',
+    'get_world_size', 'group', 'init_process_group', 'irecv',
+    'is_gloo_available', 'is_initialized', 'is_mpi_available',
+    'is_nccl_available', 'is_torchelastic_launched', 'is_ucc_available',
+    'isend', 'monitored_barrier', 'new_group', 'new_subgroups',
+    'new_subgroups_by_enumeration', 'recv', 'reduce', 'reduce_multigpu',
+    'reduce_scatter', 'reduce_scatter_multigpu', 'scatter',
+    'scatter_object_list', 'send', 'supports_complex',
```
Comment (rohan-varma) [inline torch/distributed/distributed_c10d.py:48]: We should look at these APIs and ensure that all of them are currently public and we're not introducing new public interfaces we need BC commitment to without discussion.

For example, I'm not sure if `supports_complex` needs to be public, or even things like `is_torchelastic_launched`.

File: torch/distributed/distributed_c10d.py [hunk 5]
```diff
@@ -25,16 +25,32 @@
...
+    'AllreduceCoalescedOptions', 'AllreduceOptions', 'AllToAllOptions',
+    'BarrierOptions', 'BroadcastOptions', 'GatherOptions', 'PrefixStore',
+    'ProcessGroup', 'ReduceOp', 'ReduceOptions', 'ReduceScatterOptions',
```
Comment (rohan-varma) [inline torch/distributed/distributed_c10d.py:51]: I know `dist.ProcessGroup` is technically public today, but we don't actually support this, right? All process groups should be created via init_process_Group or dist.new_group?

File: torch/distributed/distributed_c10d.py [hunk 6]
```diff
@@ -1020,7 +1067,7 @@ def send(tensor, dst, group=None, tag=0):
         group.send([tensor], group_dst_rank, tag).wait()
 
 
-def recv(tensor, src=None, group=None, tag=0):
+def recv(tensor: torch.Tensor, src: Optional[int] = None, group: Optional[ProcessGroup] = None, tag: int = 0) -> Work:
```
Comment (rohan-varma) [inline torch/distributed/distributed_c10d.py:1070]: Wonder if there's anything we can do to automatically lint that future distributed PRs introducing a new public API must be typed.

**Review by rohan-varma — COMMENTED** [review]
LGTM, but this PR seems like it could have a blast radius. Can we confirm if this PR is introducing any new public interfaces, and if so discuss those?

File: torch/distributed/__init__.py (hunk 2 above)
Comment (kumpera) [reply torch/distributed/__init__.py:76]: The issue here is that to export symbols they must belong to a module that's part of torch.dist and those are not.
This is the recommended way around to fix it.

**Pull request closed without merging**